supportedBackends/aws/src/main/scala/cromwell/backend/impl/aws/AwsBatchJob.scala

```diff
@@ -190,51 +198,47 @@ final case class AwsBatchJob(jobDescriptor: BackendJobDescriptor,           // W
-    for (inx <- 1 to 6) {
-      try {
-        if (arn == "") {
-          arn = client.registerJobDefinition(definitionRequest).jobDefinitionArn
-        }
-      } catch {
-        case e: ClientException if e.statusCode() == 404 && inx < 6 => {
-          // RegisterJobDefinition throws 404s every once in a while
-          Log.warn(s"RegisterJobDefinition 404 - retrying (${inx}/5)")
-          Thread.sleep(100L * (inx ^ 2L))
-        }
-        case e: ClientException if e.statusCode() == 409 => {
-          // This could be a problem here, as we might have registerJobDefinition
-          // but not describeJobDefinitions permissions. We've changed this to a
-          // warning as a result of this potential
-          Log.warn("Job definition already exists. Performing describe and retrieving latest revision")
-          val defs = client
-            .describeJobDefinitions(DescribeJobDefinitionsRequest.builder().jobDefinitionName(sanitize(name)).build())
-            .jobDefinitions()
-          // We need latest revision
-          defs.get(defs.size() - 1).jobDefinitionArn()
-        }
-      }
+    val submit = async.delay(client.registerJobDefinition(definitionRequest).jobDefinitionArn)
+
+    val retry: F[String] = scheduler.retry(submit, 0.millis, _ * 2, 6, {
+      case e: ClientException => e.statusCode() == 404
+      case _ => false
+    }).compile.last.map(_.get)
+
+    async.recoverWith(retry) {
+      case e: ClientException if e.statusCode() == 409 =>
+        Log.warn("Job definition already exists. Performing describe and retrieving latest revision")
+        async.
+          delay(client.describeJobDefinitions(DescribeJobDefinitionsRequest.builder().jobDefinitionName(sanitize(name).value).build())).
+          map(_.jobDefinitions()).
+          map(defs => defs.get(defs.size() - 1).jobDefinitionArn())
     }
-    arn
   }
 
+  /**
+    * This type contains only backslashes, dashes, numbers, and characters
+    */
+  type AwsStringRefinement =
+    (MatchesRegex[W.`"""[^\\\\]"""`.T] Or
+    MatchesRegex[W.`"[^A-Za-z0-9_-]"`.T]) And
+      MaxSize[W.`128`.T]
+
   /** Sanitizes a job and job definition name
-   *
-   *  @param name Job or Job definition name
-   *  @return Sanitized name
-   *
-   */
-  private def sanitize(name: String): String =
+    *
+    *  @param name Job or Job definition name
+    *  @return Sanitized name
+    *
+    */
+  private def sanitize(name: String): String Refined AwsStringRefinement =
     // Up to 128 letters (uppercase and lowercase), numbers, hyphens, and underscores are allowed.
     // We'll replace all invalid characters with an underscore
-    name
-      .replaceAll("[^A-Za-z0-9_\\-]", "_")
-      .slice(0,128)
+    refineV[AwsStringRefinement](name.replaceAll("[^A-Za-z0-9_\\-]", "_").slice(0,128)).
+      getOrElse(throw new RuntimeException("this exception is impossible to be thrown as the previous line removed all offenders"))
```

Comment:
I like to at least say `"Programmer Error! Please report this as..."` since otherwise it causes a ton of FUD when it eventually shows up (and as Jeff can attest... eventually, these always show up!)

Reply:
Sure thing